### PR TITLE
 Add Privacy and Security Considerations

### LINKED ISSUE #115: Write Privacy & Security Considerations
(A bit surprised that we don't seem to have an issue for this yet.)

For reference/inspiration we should look at concerns raised here:
- [TAG Self-Review Questionnaire: Security and Privacy](https://www.w3.org/TR/security-privacy-questionnaire/#intro)
- #113 
- #60 
- #37 
- #25

### PR DESCRIPTION
**Pull request by bvandersloot-mozilla**
Editorial change, deleting the PR template...

Hopefully this represents a good first draft. Aiming to resolve #115.

***
<a href="https://pr-preview.s3.amazonaws.com/bvandersloot-mozilla/storage-access/pull/142.html" title="Last updated on Dec 14, 2022, 2:30 PM UTC (a4ad823)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/privacycg/storage-access/142/7908c22...bvandersloot-mozilla:a4ad823.html" title="Last updated on Dec 14, 2022, 2:30 PM UTC (a4ad823)">Diff</a>

### CODE REVIEW

File: storage-access.bs [hunk 1]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
 
 <h2 id="privacy">Privacy considerations</h2>
 
-ISSUE: Write this section.
+The Storage Access API is meant able to enable the process of third-party cookie deprecation. Specifically, it enables the authenticated embed use case to continue to work, albeit under further constraints. When evaluating the privacy properties of the Storage Access API it is important to consider the pre-deprecation world of third-party cookies as the baseline. However, we aim to provide strong anti-tracking protections to users, as to not undermine the gains of third-party cookie deprecation.
+
+Importantly, we do not degrade privacy properties when compared to pre-deprecation of third-party cookies. This follows from a lack of platform-specific information used in the spec to prevent stateless tracking and the only state added being a permission scoped to the [=site|sites=] of the embedding and embedded [=Document=].
```
Comment (johannhof) [inline storage-access.bs:315]: This is okay for now but it might be better to factor this point out into a separate header, such as "Do not introduce additional cross-site state or fingerprinting surfaces". It might make sense to just create this header even if we just have one paragraph about it.

File: storage-access.bs [hunk 2]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
+Our privacy considerations are more challenging where default third-party cookies are already deprecated. The challenge is to decide when and how to permit the Storage Access API to be used to revert a cookie-less (or cookie-partitioned) iframe to a pre-deprecation state, giving it access to its [=unpartitioned data=].
```
Comment (johannhof) [inline storage-access.bs:317]: I think I'd remove the first sentence here and add a new header along the lines of "Gating storage access on trusted user signals" above it.

File: storage-access.bs [hunk 3]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
+In an ideal case, an iframe would only be able to gain access to its [=unpartitioned data=] if:
```
Comment (johannhof) [inline storage-access.bs:319]: I'm not sure if we should mirror the required steps in the algorithm here, we should try to have a single source of truth and just link to the steps in the determine storage access algorithm.

File: storage-access.bs [hunk 4]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
+1. the user interacts with the iframe, giving it [=transient activation=]
+2. the iframe is permitted by the embedder to use the API
+3. the iframe is a [=secure context=]
+4. the user grants express, pairwise permission to the embeddee to use its cookies in the embedder
+5. the user is not inundated by requests for the "<code>storage-access</code>" permission so their express permission is not undermined by fatigue
+
+This specification requires the first three of implementers. This provides guarantees that the user is aware of the content of the iframe, the embedder has not opted out of the iframe's authentication, and the third-party cookies are not disclosed to network attackers, respectively.
+
+The last two points are in tension. In an ideal world, we would show a prompt to the user in every call to {{Document/requestStorageAccess()}}. But, this would allow pages to prompt the user so frequently as to put the last point at the discretion of the page– a state we find unacceptable.
```
Comment (johannhof) [inline storage-access.bs:329]: I think that it's great to mention that user agents have a lot of freedom around prompting and permissions due to the implementation-defined steps. However, could we slightly rephrase this to give guidance to user agents instead?

File: storage-access.bs [hunk 5]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
 <figcaption>An example prompt which could be shown to the user when a site calls `document.`{{Document/requestStorageAccess()}}.</figcaption>
 </figure>
 
+Thus, the last two points represent a key point of compromise. We permit [=implementation-defined=] behavior on when to grant or deny requests for [=unpartitioned data=] without requiring user choice so long as they meet all other requirements. This compromise weakens the privacy guarantees of this proposal, specifically point 4, to a degree under the control of the implementer and gives the implementer the power to render it impossible to get storage access with this API. However, this has proven necessary to enable condition 5 to be possible given our implementers' differing stances on the compromise between these two points.
+
+An important avenue for improvement is the removal of the [=implementation-defined=] silent grants and denies. This would improve the developer experience using the API across platforms.
```
Comment (johannhof) [inline storage-access.bs:339]: I don't think we need to discuss future improvements to the spec here, but not a strong opinion. Again, I think we should rather point to those compromises and advise user agents on how to handle them best.

File: storage-access.bs [hunk 6]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
+Another tension in the design of the API is what to use to key the "<code>storage-access</code>" permission: [=url/origin=] or [=site|sites=]. We chose [=site|sites=] because we believe them to be acceptable boundaries for privacy while enabling existing uses of [=same site=] and cross-origin iframes on the same page with only one user prompt.
```
Comment (johannhof) [inline storage-access.bs:341]: I think this should be a separate subsection under a new header.

File: storage-access.bs [hunk 7]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
 <h2 id="security">Security considerations</h2>
 
-ISSUE: Write this section.
+It is important that this spec not degrade security properties of the web platform, even when compared to post-deprecation of third-party cookies. Third-party cookie deprecation has potential benefits for security, specifically in mitigating attacks that rely upon authenticated requests, e.g. CSRF. We do not wish the Storage Access API to be a foothold for such attacks to leverage.
+
+To this end, we limit the impact of a "<code>storage-access</code>" permission grant to only give access to [=unpartitioned data=] to the iframe that called {{Document/requestStorageAccess()}} and only until the iframe navigates across an [=url/origin=] boundary. This ensures that only [=url/origin||origins=] with a page that call {{Document/requestStorageAccess()}} will be making credentialed requests, and moreover the embedee page can control which embedder it permits via the Content Security Policy "<code>frame-ancestors</code>" directive. This retains an [=url/origin=]-scoped control for security purposes by the embedee.
```
Comment (johannhof) [inline storage-access.bs:347]: Should those origin references just point to [=origin=]?

File: storage-access.bs [hunk 8]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
+This also is effective at preventing another attack: one on the embedee's reputation. We consider any third-party authenticated request to have potential reputational harm as consumers become more privacy conscious. Therefore a first-party or sibling third-party causing an embedded resource to be requested with the user's authentication cookies would constitute an attack on the reputation of that third-party's owner. This is also a reason we require this API to be used in a [=secure context=]: so a network adversary cannot induce an embedee to use this API.
```
Comment (johannhof) [inline storage-access.bs:349]: Should be a separate subsection :)

File: storage-access.bs [hunk 9]
```diff
@@ -310,17 +310,47 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
...
+
+The embedder has control over which iframes have the ability to become authenticated, or even display a permission request to the user via the Permission Policy and iframe sandboxing.
+
+We also consider notification abuse permitted by the embedder to be in-scope. Specifically, we require user interaction in the iframe and consume that rejection on a denial to restrict the conditions a permission prompt will be shown to the user. This attacks such as re-requesting a permission immediately after the user denies it.
```
Comment (johannhof) [inline storage-access.bs:353]: Should be a separate subsection. Again, I think it's fine to just add a header and have this sentence there for future improvement.

Comment (johannhof) [inline storage-access.bs:353]: Should we drop the "we also consider ... to be in-scope"? In scope of what?

Comment (johannhof) [inline storage-access.bs:353]:
```suggestion
We also consider notification abuse permitted by the embedder to be in-scope. Specifically, we require user interaction in the iframe and consume that rejection on a denial to restrict the conditions a permission prompt will be shown to the user. This mitigates attacks such as re-requesting a permission immediately after the user denies it.
```

**Review by johannhof — CHANGES_REQUESTED** [review]
Thanks for giving this a stab, Ben! I had a couple of comments.

I still think we could add some more content in the future, but we shouldn't iterate on this PR forever. So I'm happy to merge once those comments are fixed :)

File: storage-access.bs (hunk 7 above)
Comment (bvandersloot-mozilla) [reply storage-access.bs:347]: I get the following error from BS with that:
```
LINE ~346: Multiple possible 'origin' dfn refs.
Arbitrarily chose https://dom.spec.whatwg.org/#concept-document-origin
To auto-select one of the following refs, insert one of these lines into a <pre class=link-defaults> block:
spec:dom; type:dfn; text:origin
spec:fetch; type:dfn; text:origin
spec:html; type:dfn; for:/; text:origin
spec:html; type:dfn; for:environment settings object; text:origin
spec:service-workers; type:dfn; text:origin
spec:url; type:dfn; text:origin
[=origin=]
```

File: storage-access.bs (hunk 3 above)
Comment (bvandersloot-mozilla) [reply storage-access.bs:319]: I think keeping abstract points here may be useful because they can guide iteration on the algorithms. I do share your concern about having more than one source of truth.

File: storage-access.bs (hunk 7 above)
Comment (johannhof) [reply storage-access.bs:347]: Ah, yes, try `[=/origin=]`

File: storage-access.bs (hunk 4 above)
Comment (bvandersloot-mozilla) [reply storage-access.bs:329]: I'll add guidance to the end of the paragraph.

File: storage-access.bs (hunk 5 above)
Comment (bvandersloot-mozilla) [reply storage-access.bs:339]: I'm willing to drop it, but do think to brings up a developer experience loss that was made with this compromise.

File: storage-access.bs (hunk 9 above)
Comment (bvandersloot-mozilla) [reply storage-access.bs:353]: Re-worded. I was thinking in-scope of the tradeoffs considered... but that is really confusing and doesn't make much sense.

File: storage-access.bs [hunk 10]
```diff
@@ -272,9 +270,9 @@ ISSUE: [since this is UA-defined, does it make sense to follow-up separately wit
 
 <h3 id="navigation">Changes to navigation</h3>
 
-Before changing the [=current entry=] of a [=session history=], run the following steps:
+Before changing the current entry of a session history, run the following steps:
 
-1. Let |doc| be [=current entry=]'s {{Document}}.
```
[the remark this answers is not in the document]
Comment (bvandersloot-mozilla) [reply storage-access.bs:277]: I have no idea, but let me know if they get in the way of merging. I can rebase to try to remove them.

File: storage-access.bs (hunk 7 above)
Comment (bvandersloot-mozilla) [reply storage-access.bs]: That works perfectly. Thank you for the help with the dark magic of Bikeshed

File: storage-access.bs [hunk 11]
```diff
@@ -20,8 +20,6 @@ Complain About: accidental-2119 true
 
 <!-- File issues on HTML to export each of these -->
 <pre class=link-defaults>
-spec:html; type:dfn; text:session history; url:https://html.spec.whatwg.org/multipage/history.html#session-history
-spec:html; type:dfn; text:current entry; url:https://html.spec.whatwg.org/multipage/history.html#current-entry
```
Comment (annevk) [inline storage-access.bs:24]: \o/

File: storage-access.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (annevk) [reply storage-access.bs:277]: I think that's because it's not rebased against main, but there are no conflicts as these are the same changes.

Squashing and rebasing against main might be slightly safer.

File: storage-access.bs [hunk 12]
```diff
@@ -310,17 +310,52 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
 
 <h2 id="privacy">Privacy considerations</h2>
 
-ISSUE: Write this section.
+The Storage Access API is meant able to enable the deprecation of third-party cookies. Specifically, it allows the authenticated embeds use case to continue to work. As such, the API provides a way for developers to re-gain access to third-party cookies, albeit under further constraints. While that comes with a risk of abuse by third parties for tracking purposes, it is an explicit goal of the API and a key to its design to not undermine the gains of third-party cookie deprecation.
```
Comment (annevk) [inline storage-access.bs:313]: "cross-site cookies", right? Avoiding "third party" in the specification is probably a good idea given https://tess.oconnor.cx/2020/10/parties.

Comment (annevk) [reply storage-access.bs:313]: Also:

* removal, not deprecation
* I think it would be good to flush this out a bit with a more worked out example of https://website.example and https://authenticated-widget.example and how cookies end up being impacted here. (And perhaps also stipulate these are the same cookies as you'd get with a direct navigation to the embed?)

File: storage-access.bs (hunk 3 above)
Comment (annevk) [reply storage-access.bs:318]: I think we should talk about a nested document here. "iframe" isn't really a term-of-art and also doesn't hold the authority.

File: storage-access.bs (hunk 5 above)
Comment (annevk) [reply storage-access.bs:338]: Phrasing this as an encouragement on user agents to do better on this makes more sense to me as well.

**Review by annevk — COMMENTED** [review]
I think it's worth thinking about the end state and writing this section from that perspective. I.e., assume cross-site cookie removal happened. And perhaps highlight issues if that hasn't happened yet. Ideally we don't have to rewrite things too much once that has happened everywhere, though removing certain bits seems okay.

File: storage-access.bs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bvandersloot-mozilla) [reply storage-access.bs]: Resolved.

File: storage-access.bs [hunk 13]
```diff
@@ -310,17 +310,56 @@ The Storage Access API defines a [=policy-controlled feature=] identified by the
 
 <h2 id="privacy">Privacy considerations</h2>
 
-ISSUE: Write this section.
+The Storage Access API enabled the removal of cross-site cookies. Specifically, it allows the authenticated embeds use case to continue to work. As such, the API provides a way for developers to re-gain access to cross-site cookies, albeit under further constraints.
```
Comment (johannhof) [inline storage-access.bs:313]:
```suggestion
The Storage Access API enables the removal of cross-site cookies. Specifically, it allows the authenticated embeds use case to continue to work. As such, the API provides a way for developers to re-gain access to cross-site cookies, albeit under further constraints.
```

**Review by johannhof — APPROVED** [review]
Thanks Ben! I'll merge this, and let's keep iterating on these as needed going forward.
